src/ucs/arch/x86_64/cpu.c

```diff
@@ -94,37 +99,57 @@ static double ucs_x86_tsc_freq_from_cpu_model()
         max_ghz = ucs_max(max_ghz, ghz);
         if (max_ghz != ghz) {
             warn = 1;
+            break;
         }
     }
     fclose(f);
 
     if (warn) {
-        ucs_debug("Conflicting CPU frequencies detected, using fallback timer");
-        ucs_arch_x86_enable_rdtsc = 0;
+        level = ucs_global_opts.warn_inv_tsc ? UCS_LOG_LEVEL_WARN :
+                                               UCS_LOG_LEVEL_DEBUG;
+        ucs_log(level,
+                "Conflicting CPU frequencies detected, using fallback timer");
+
+        return -1;
     }
+
     return max_ghz * 1e9;
 }
 
-double ucs_arch_get_clocks_per_sec()
+double ucs_x86_init_tsc_freq()
 {
     double result;
 
-    ucs_x86_check_invariant_tsc();
+    if (!ucs_x86_invariant_tsc()) {
+        goto err_disable_rdtsc;
+    }
 
-    if (ucs_arch_x86_enable_rdtsc) {
-        /* First, try to find the information in the model name string (will work on Intel) */
-        result = ucs_x86_tsc_freq_from_cpu_model();
-        if (result == 0) {
-            /* Read clock speed from cpuinfo */
-            result = ucs_get_cpuinfo_clock_freq("cpu MHz", 1e6);
-        }
-        if ((result > 0.0) && ucs_arch_x86_enable_rdtsc) {
-            /* using rdtsc */
-            return result;
-        }
+    ucs_arch_x86_enable_rdtsc = UCS_YES;
+
+    result = ucs_x86_tsc_freq_from_cpu_model();
+    if (result <= 0.0) {
+        result = ucs_get_cpuinfo_clock_freq("cpu MHz", 1e6);
+    }
+
+    if (result > 0.0) {
+        return result;
+    }
+
+err_disable_rdtsc:
+    ucs_arch_x86_enable_rdtsc = UCS_NO;
+    return -1;
+}
+
+double ucs_arch_get_clocks_per_sec()
```

Comment:
maybe unite ucs_x86_init_tsc_freq() and ucs_arch_get_clocks_per_sec() to one function?

Reply:
I'd leave it as it is, because
- `ucs_arch_get_clocks_per_sec()` is a common function for different archs,  
- `ucs_x86_init_tsc_freq()` is a initialization part, which is used for getting time and we do not need to call  `ucs_arch_generic_get_clocks_per_sec` for it

Reply:
so, in `ucs_arch_get_clocks_per_sec()`, maybe better check the value of `ucs_arch_x86_enable_rdtsc` to in line 148 instead of 'freq' ?

Reply:
but we have to get frequency in `ucs_arch_get_clocks_per_sec`, not the state of rdtsc

Reply:
return frequency, but use state of rdtsc to decide whether to call `ucs_arch_generic_get_clocks_per_sec()`